### ISSUE TITLE
Custom Dataset Creation and Model Training Pipeline

### ISSUE THREAD

**Issue by shihao1989** [shihao1989/lidar-p1-port-monitoring#15]
---
name: Custom Dataset Creation and Model Training Pipeline
status: open
created: 2025-09-13T12:09:32Z
updated: 2025-09-15T12:00:00Z
github: https://github.com/shihao1989/lidar-p1-port-monitoring/issues/15
depends_on: [13, 14]
parallel: false
conflicts_with: []
---

# Task: Custom Dataset Creation and Model Training Pipeline

## Description
Deploy comprehensive annotation ecosystem using mature open-source tools (LATTE + 3D-BAT 2.0 + labelCloud + Semantic Editor) with MMDetection3D v1.4+ training pipeline for RTX 4090-optimized port container detection, achieving 6.2x annotation efficiency improvement through tool integration.

### **Annotation Tools Ecosystem (Production-Ready)**
- **LATTE** (UC Berkeley, 6.2x efficiency boost)
  - Repository: `github.com/bernwang/latte`
  - Features: Sensor fusion, one-click annotation, tracking
  - Performance: 23.6% higher instance precision, 2.2% higher recall
- **3D-BAT 2.0** (Active learning support, 2024/01 update)
  - Repository: `github.com/walzimmer/bat-3d`
  - Features: Web-based collaboration, full-surround data streams
  - Updates: Active learning integration, improved multi-modal support
- **labelCloud** (Lightweight desktop, 2024 updates)
  - Repository: `github.com/ch-sa/labelCloud`
  - Features: Semantic segmentation mode, multi-format support
  - Formats: PLY, PCD, PTS, XYZ, BIN (Python 3.7-3.9)
- **Semantic Segmentation Editor** (Hitachi, web-based)
  - Repository: `github.com/Hitachi-Automotive-And-Industry-Lab/semantic-segmentation-editor`
  - Features: Point cloud semantic labeling, 3D object creation

### **MMDetection3D Training Stack**
- **MMDetection3D v1.4+** (Latest stable with BEVFusion)
  - Enhanced TensorRT ≥8.6 support for dynamic shapes
  - Port-specific anchor optimization
  - Custom dataset integration patterns
- **Training Optimization for RTX 4090**
  - Single GPU focus (multi-GPU scaling issues)
  - 24GB VRAM optimization strategies
  - TF32 training: 1.3x-1.9x speedup vs RTX 3090
  - FP16 training: 1.3x-1.8x speedup vs RTX 3090

### **Dataset Format Standardization**
```
Port Container Dataset (MMDetection3D Compatible)
├── KITTI Structure: /data/port_container/
│   ├── training/velodyne/    # Point clouds (.bin)
│   ├── training/label_2/     # 3D boxes (.txt)
│   ├── training/calib/       # Calibration (.txt)
│   └── training/image_2/     # Camera images (.png)
└── Semantic Structure: /data/port_semantic/
    ├── sequences/XX/velodyne/  # Point clouds
    ├── sequences/XX/labels/    # Semantic labels
    └── semantic-kitti.yaml    # 9-class port mapping
```

### **Multi-Tool Workflow Integration**
```
LATTE (6.2x speed) → 3D Detection Boxes → KITTI Format
     ↓
3D-BAT 2.0 (Collaboration) → Quality Review → Refinement
     ↓
labelCloud (Desktop) → Semantic Segmentation → Point Labels
     ↓
Semantic Editor (Web) → Final QC → MMDetection3D Dataset
```

### Multi-Tool Annotation Pipeline
- [ ] LATTE annotation environment deployed on RTX 4090 workstation (6.2x efficiency improvement)
- [ ] 3D-BAT 2.0 web-based annotation platform for collaborative labeling
- [ ] labelCloud lightweight tool for 3D bounding box annotation
- [ ] UNA-PCL web-based tool for point cloud segmentation tasks
- [ ] Integrated workflow supporting both segmentation and 3D detection annotation

### MMDetection3D Integration
- [ ] KITTI-format dataset structure established for MMDetection3D compatibility
- [ ] Custom PortDataset class implementation with 5-class definitions
- [ ] PointPillars model configuration optimized for port environment
- [ ] CenterPoint model configuration for high-accuracy detection
- [ ] Custom anchor sizes and point cloud ranges for port-specific objects

### Advanced Annotation Features
- [ ] Semi-automatic annotation pipeline with Mask R-CNN pre-labeling integration
- [ ] Point cloud segmentation workflow for object boundary refinement
- [ ] One-click 3D bounding box generation using sensor fusion (LiDAR + camera)
- [ ] Sequence tracking and automatic label propagation across frames
- [ ] 5-class labeling system: Person, Vehicle, Container, Crane, Forklift with standardized dimensions

### Semantic Segmentation Dataset Creation
- [ ] Ground segmentation annotation for port environment (flat yards, ramps, uneven terrain)
- [ ] Dense container stacking semantic labels for overlapping detection scenarios
- [ ] Patchwork++ ground segmentation integration for automated ground truth generation
- [ ] 2DPASS semantic segmentation model training dataset with port-specific classes
- [ ] LDLS-compatible 2D-3D fusion annotation pipeline for container instance segmentation
- [ ] Point-wise semantic labels: Ground, Container, Person, Vehicle, Crane, Forklift, Infrastructure

### Quality Control & Dataset Management
- [ ] Quality control system with 3-level validation (auto, peer-review, expert)
- [ ] Dataset with 10,000+ high-quality annotated scenes using sensor fusion approach
- [ ] Semantic segmentation quality metrics: IoU, pixel accuracy, boundary precision
- [ ] Data augmentation pipeline optimized for port environment (sea wind, lighting variations)
- [ ] Batch processing tools for handling large-scale annotation projects
- [ ] Standardized annotation guidelines for both detection and segmentation tasks

### Implementation Approach
- **Multi-tool annotation workflow**: LATTE + 3D-BAT 2.0 + labelCloud + UNA-PCL integration
- **MMDetection3D optimization**: PointPillars and CenterPoint models for RTX 4090
- **Segmentation + Detection**: Combined point cloud segmentation and 3D bounding box labeling
- **Dataset format standardization**: KITTI-compatible with MMDetection3D custom config

### Key Technical Considerations
- **Port-specific model optimization**: Custom anchor sizes and point cloud ranges
- **Multi-modal annotation**: LiDAR point clouds with camera image fusion
- **RTX 4090 VRAM optimization**: Memory-efficient processing for large point clouds
- **Quality control automation**: 3-level validation with automated consistency checks
- **Semantic segmentation integration**: Patchwork++ ground segmentation and 2DPASS models
- **Dense container detection**: LDLS 2D-3D fusion for overlapping container instance segmentation

### Code Locations
- `/tools/annotation/latte/` - LATTE annotation environment setup
- `/tools/annotation/3d-bat/` - 3D-BAT 2.0 web platform deployment
- `/tools/annotation/labelcloud/` - labelCloud desktop tool configuration
- `/tools/annotation/una-pcl/` - UNA-PCL web-based segmentation
- `/tools/segmentation/patchwork++/` - Ground segmentation algorithm integration
- `/tools/segmentation/2dpass/` - Semantic segmentation model training
- `/tools/segmentation/ldls/` - Container instance segmentation pipeline
- `/data/datasets/port_container/` - MMDetection3D KITTI structure
- `/data/datasets/port_semantic/` - Semantic segmentation dataset structure
- `/configs/port/` - Custom model configurations for port environment

### Dataset Format Specifications

**MMDetection3D KITTI Structure:**
```
data/port_container/
├── ImageSets/
│   ├── train.txt        # Training file indices
│   └── val.txt          # Validation file indices
├── training/
│   ├── calib/           # Calibration files (.txt)
│   ├── image_2/         # Camera images (.png)
│   ├── label_2/         # 3D bounding box labels (.txt)
│   ├── velodyne/        # Point clouds (.bin format)
│   └── planes/          # Ground plane estimates (.txt)
└── testing/
    ├── calib/
    ├── image_2/
    └── velodyne/
```

**Semantic Segmentation Dataset Structure:**
```
data/port_semantic/
├── sequences/
│   ├── 00/
│   │   ├── velodyne/    # Point clouds (.bin format)
│   │   ├── labels/      # Point-wise semantic labels (.label)
│   │   └── poses.txt    # Sensor poses for sequence
│   ├── 01/
│   └── ...
├── semantic-kitti.yaml  # Class definitions and color mapping
└── train_val_split.txt  # Training/validation sequence split
```

**3D Detection Label Format (MMDetection3D Compatible):**
```
# Format: [class_name score x y z dx dy dz yaw]
Person 0.99 1.23 1.42 0.23 0.6 0.6 1.8 1.56
Vehicle 0.95 3.51 2.15 0.42 4.5 2.0 1.8 1.23
Container 0.98 5.67 -1.20 0.85 12.2 2.4 2.6 0.78
Crane 0.89 -2.1 4.5 2.1 15.0 8.0 25.0 -0.45
Forklift 0.94 0.95 -3.2 0.65 3.0 1.5 2.5 1.12
```

**Semantic Segmentation Label Format (.label files):**
```python
# Point-wise semantic labels (uint32 format)
# Label mapping for port environment:
semantic_labels = {
    0: 'unlabeled',      # Unknown/noise points
    1: 'ground',         # Port yard ground, roads, ramps
    2: 'container',      # 20ft/40ft shipping containers
    3: 'person',         # Dock workers, inspectors
    4: 'vehicle',        # Trucks, inspection vehicles
    5: 'crane',          # Gantry cranes, mobile cranes
    6: 'forklift',       # Container handling equipment
    7: 'infrastructure', # Buildings, fences, bollards
    8: 'water',          # Harbor water surface
    9: 'vegetation'      # Trees, grass in port areas
}

# Instance segmentation format (for dense containers)
instance_labels = {
    'container_id': [semantic_id, instance_id],  # [2, unique_id]
    'stacking_level': int,                       # Container stack position
    'container_type': str                        # '20ft' or '40ft'
}
```

**Class-Specific Dimensions (Port Environment):**
```python
# MMDetection3D anchor configuration
port_anchor_sizes = {
    'Person': [0.6, 0.6, 1.8],      # Dock workers, inspectors
    'Vehicle': [4.5, 2.0, 1.8],     # Trucks, inspection vehicles
    'Container': [12.2, 2.4, 2.6],  # 40ft container (20ft: [6.1, 2.4, 2.6])
    'Crane': [15.0, 8.0, 25.0],     # Gantry cranes (highly variable)
    'Forklift': [3.0, 1.5, 2.5]     # Container handling equipment
}

# Point cloud range for port environment
point_cloud_range = [0, -50, -3, 100, 50, 15]  # [x_min, y_min, z_min, x_max, y_max, z_max]
```

### Open-Source Annotation Tools Integration

**1. LATTE (Semi-Automatic 3D Annotation)**
- Repository: https://github.com/bernwang/latte
- Features: Sensor fusion, automatic pre-labeling, RTX 4090 optimization
- Use case: Primary annotation tool with 6.2x efficiency improvement

**2. 3D-BAT 2.0 (Web-Based Collaborative)**
- Repository: https://github.com/walzimmer/bat-3d
- Features: Web interface, multi-user collaboration, LiDAR+camera fusion
- Use case: Team annotation workflow, remote collaboration

**3. labelCloud (Lightweight Desktop)**
- Repository: https://github.com/ch-sa/labelCloud
- Features: Simple 3D bounding box annotation, various format support
- Use case: Quick annotation, validation, and correction tasks

**4. UNA-PCL (Web Segmentation)**
- Repository: https://github.com/gengarneko/una-pcl
- Features: Vue.js + Three.js, point cloud segmentation, efficient labeling
- Use case: Detailed segmentation for object boundary refinement

### Semantic Segmentation Tools Integration

**5. Patchwork++ (Ground Segmentation)**
- Repository: https://github.com/url-kaist/patchwork-plusplus
- Features: Real-time ground segmentation, terrain-robust, no parameter tuning
- Use case: Automated ground truth generation for port terrain (flat yards, ramps)

**6. 2DPASS (LiDAR Semantic Segmentation)**
- Repository: https://github.com/yanx27/2DPASS
- Features: 2D-3D fusion, multi-scale learning, port-adaptable architecture
- Use case: Point-wise semantic segmentation model training for port scenarios

**7. LDLS (Dense Container Detection)**
- Repository: https://github.com/PRBonn/LDLS
- Features: 2D-3D fusion, instance segmentation, overlapping object detection
- Use case: Dense container stacking detection with instance-level precision

**8. labelCloud Semantic Extension**
- Repository: https://github.com/ch-sa/labelCloud (with semantic plugin)
- Features: Point-wise labeling, semantic class assignment, instance ID tracking
- Use case: Manual correction of automated semantic segmentation results

### Hardware Requirements
- [ ] Task #3 completion (calibrated sensors for accurate ground truth)
- [ ] Task #4 completion (RTX 4090 data collection infrastructure)
- [ ] RTX 4090 workstation with 64GB RAM for large point cloud processing
- [ ] High-resolution dual monitors for efficient annotation workflow

### Software Infrastructure
- [ ] LATTE installation and RTX 4090 GPU acceleration setup
- [ ] 3D-BAT 2.0 web platform deployment and configuration
- [ ] labelCloud desktop application setup across annotation workstations
- [ ] UNA-PCL web-based tool deployment for segmentation tasks
- [ ] Patchwork++ ground segmentation algorithm integration
- [ ] 2DPASS semantic segmentation model training environment
- [ ] LDLS container instance segmentation pipeline deployment
- [ ] MMDetection3D framework installation with custom port dataset support

### Team & Training
- [ ] Data annotation team recruitment and training program
- [ ] Port-specific labeling standards and guidelines development
- [ ] Multi-tool workflow training for annotation team
- [ ] Quality control reviewer training and certification

## Effort Estimate
- Size: XXL (expanded from XL due to multi-tool integration)
- Hours: 136 (80 original + 56 for annotation system integration)
- Parallel: false (requires calibration and data pipeline)

### Model Performance & Deployment
- [ ] PointPillars model achieving >85% mAP on port validation dataset
- [ ] CenterPoint model achieving >90% mAP on port validation dataset
- [ ] 2DPASS semantic segmentation achieving >75% mIoU on port semantic dataset
- [ ] Patchwork++ ground segmentation achieving >95% precision on port terrain
- [ ] LDLS container instance segmentation achieving >80% AP for overlapping containers
- [ ] Training pipeline producing models meeting >95% accuracy requirements
- [ ] TensorRT optimization for RTX 4090 inference deployment
- [ ] Automated model evaluation and performance reporting

### Annotation System Validation
- [ ] All 8 annotation tools (4 detection + 4 segmentation) fully operational with RTX 4090 optimization
- [ ] Multi-tool workflow validated with 100+ annotated frames (detection + segmentation)
- [ ] Quality control pipeline achieving >95% annotation accuracy for both tasks
- [ ] Semantic segmentation annotation achieving >90% point-wise accuracy
- [ ] Annotation team trained and productive (100-150 frames/person/day)
- [ ] Standardized workflows documented and validated for both detection and segmentation

### Integration & Documentation
- [ ] Integration tests with real-time inference pipeline (detection + segmentation)
- [ ] Model deployment artifacts and version control for all model types
- [ ] MMDetection3D dataset format validation and compatibility tests
- [ ] Semantic segmentation dataset format validation (SemanticKITTI compatible)
- [ ] Documentation for dataset expansion and model retraining (both modalities)
- [ ] Comprehensive annotation guidelines and tool usage documentation for 8-tool workflow
